src/pudl/extract/extractor.py

```diff
@@ -0,0 +1,151 @@
+"""Generic functionality for extractors."""
+from collections import defaultdict
+
+import pandas as pd
+from dagster import (
+    AssetsDefinition,
+    DynamicOut,
+    DynamicOutput,
+    OpDefinition,
+    graph_asset,
+    op,
+)
+
+
+# TODO: Consider abstract class
+class GenericExtractor:
```

Comment:
I created this module to capture a number of top-level functions (below) that were no longer scoped to just Excel extractors, given how we want to pass multiple years to extract CSV data for. To get the inheritance hierarchy and typing represented so far to work, I created this stub base class and plan to look into the proper way to make this abstract, or Python's equivalent of an interface.

Reply:
I think you've got a choice of using a `Protocol` (Python's non-binding answer to interfaces / structural typing) or an abstract base class (`abc.ABC`) and some `@abstractmethod` decorators. We've mostly gone with ABC elsewhere.

I liked this video exploring the two options:
https://www.youtube.com/watch?v=xvb5hGLoK0A